akka-http2-support/src/main/scala/akka/http/impl/engine/http2/AlpnSwitch.scala

```diff
@@ -89,28 +100,59 @@ object AlpnSwitch {
                   propagatePull
                 case Some(ele) ⇒
                   new OutHandler {
-                    def onPull(): Unit = {
+                    override def onPull(): Unit = {
                       out.push(initialElement.get)
                       out.setHandler(propagatePull)
                     }
+
+                    override def onDownstreamFinish(): Unit = {
+                      out.complete()
+                      super.onDownstreamFinish()
+                    }
                   }
                 case None ⇒ propagatePull
               }
 
             out.setHandler(firstHandler)
             setHandler(in, new InHandler {
-              def onPush(): Unit = out.push(grab(in))
+              override def onPush(): Unit = out.push(grab(in))
+
+              override def onUpstreamFinish(): Unit = {
+                out.complete()
+                super.onUpstreamFinish()
+              }
+
+              override def onUpstreamFailure(ex: Throwable): Unit = {
+                out.fail(ex)
+                super.onUpstreamFailure(ex)
+              }
             })
 
             if (out.isAvailable) pull(in) // to account for lost pulls during initialization
           }
           def connect[T](in: SubSinkInlet[T], out: Outlet[T]): Unit = {
-            in.setHandler(new InHandler {
-              def onPush(): Unit = push(out, in.grab())
-            })
-            setHandler(out, new OutHandler {
-              def onPull(): Unit = in.pull()
-            })
+            val handler = new InHandler {
+              override def onPush(): Unit = push(out, in.grab())
+
+              override def onUpstreamFinish(): Unit = {
+                in.cancel()
+                super.onUpstreamFinish()
+              }
+              override def onUpstreamFailure(ex: Throwable): Unit = {
+                in.cancel()
+                super.onUpstreamFailure(ex)
+              }
+            }
+
+            val outHandler = new OutHandler {
+              override def onPull(): Unit = in.pull()
+              override def onDownstreamFinish(): Unit = {
+                in.cancel()
```

Comment:
Yes, we need both, otherwise, we might still keep the `SubSource` / `SubSink` alive depending on how the race between completion / cancellation ends up.